### DISCUSSION TITLE
Issue running openfast simulation interfaced with ROSCO Controller

### DISCUSSION THREAD

**Discussion by ema-chi** [NatLabRockies/ROSCO#201] [Q&A]
Hi, i'm trying to run openfast simulation interfaced with ROSCO Controller but i encounter an error, these are the steps i've done:

- built openfast v2.4.0 (i'm using this version for personal purpose) obtaining build/bin folder
- created S-Function with create_FAST_SFunc.m file
- changed the file paths of runFast.m (inside ROSCO-raaw1.3\Matlab_Toolbox  path)  with the correct ones

During the first run i encountered this error:

"Error using runFAST
The expression: u(strmatch('GenSpeed',OutList))
in 'ROSCO/Extract Signals/Extract GenSpeed'
has a syntax error"

Looking for it I found that it was a problem related to the use of strmatch in the "Fcn" block, so i changed them with "Interpreted Matlab Function" blocks.

Next error was:

"Error using runFAST
Error reported by S-function 'FAST_SFunc' in 'ROSCO/FAST Nonlinear Wind Turbine/S-Function':
FAST_InitializeAll:FAST_Init:FAST_ReadPrimaryFile:Invalid logical input for file
"C:\...\Controller_MATLAB_interface_debug\ROSCO-raaw1.3\ROSCO-raaw1.3\Test_Cases\5MW_Land_Simulink\5MW_Land_Simulink.fst"
occurred while trying to read SumPrint."

I checked this out and i found out it was related with different version of input files, so i changed them according to v2.4.0 format.

Last error is:

"Error using runFAST
Error reported by S-function 'FAST_SFunc' in 'ROSCO/FAST Nonlinear Wind Turbine/S-Function':
FAST_Update:size of InputAry is invalid."

I've searched for it but i couldn't find anything about it.
Thanks to anyone who can help.

I'm using:

- Microsoft Visual Studio 2019
- Matlab R2022b

**Comment by dzalkind** [comment]
The inputs to the SFunction also changed from OpenFAST v2.4.0 to the present.  I would recommend looking [here](https://github.com/OpenFAST/openfast/blob/v2.4.0/glue-codes/simulink/examples/OpenLoop.mdl) and reproducing that in your setup.

**Reply by ema-chi** [reply]
@dzalkind  thank you for support!
I'm not that familiar with openfast, i made changes and the code run correctly, but i'd like to know if i'm doing right.

These are the inputs of S-Function in OpenLoop.mdl example model:

![Cattura](https://user-images.githubusercontent.com/121025653/209313980-029abae5-1250-4f50-aed5-299b044b5969.PNG)

So i changed the inputs in ROSCO Controller's S-Function according to those, commenting out the last three inputs that are not present in OpenLoop model:

![Screenshot (79)](https://user-images.githubusercontent.com/121025653/209314405-e8bb30de-e8a9-4be6-a5fa-5f7b341c2ce5.jpg)

I also modified the third input of S-Function block: initially it was "43", but it gave me error and changing it with "0" the code runs without errors.
From the Openfast's guide, i read this third parameter refers to **NumAdditionalInputs** and that it's "0" for most cases.

I'd really appreciate if you could tell me if i misunderstood your reply or if i'm making mistakes.

Thank you in advance!

**Reply by dzalkind** [reply]
If it's working, then I think you are okay.  I would run a couple of cases and see if the outputs are plausible.

**Reply by ema-chi** [reply]
@dzalkind thank you, i'm going to follow your suggests.

**Comment by severalautumn** [comment]
Hi, ema-chi!

Thank you so much for your help! I encountered the same problem as you.

The problem related to the use of strmatch in the "Fcn" block was solved by your method.

About the second error, which is related to different versions of input files, could you please provide more details on how to change them according to v2.4.0?

Thanks in advance for your help.

**Reply by ema-chi** [reply]
Hi @severalautumn,

I'm glad this helped you! 

Regarding the second error, are you having problems using ROSCO files with the v2.4 exe? 
Anyway, unfortunately some time has passed and i don't remember exactly what i've done. 
It should be about differences between ROSCO input files version and the v2.4 accepted format files. I suggest you to compare .IN and .fst files between ROSCO and a v2.4 files to check if there exists some differences, then changing the ROSCO files the same way the v2.4 are made.

**Reply by severalautumn** [reply]
Hi @ema-chi ,

Thanks for your so prompt and kind reply!

I have got the ROSCO v2.4.1 running successfully on the MATLAB/Simulink platform.

This error was encountered when I tried the latest version 2.7.0 of ROSCO.

![1679662886786](https://user-images.githubusercontent.com/128788048/227528121-ac2c655a-5e55-4885-b6f5-e1c20a6fa851.png)


I will try checking the .IN and .fst files as you suggested.

Thanks again for your generous help.
